flink-core/src/main/java/org/apache/flink/core/fs/LimitedConnectionsFileSystem.java

```diff
@@ -0,0 +1,1097 @@
+				throw exception;
+			} else {
+				// throw a new exception to capture this call's stack trace
+				throw new StreamTimeoutException(timeoutException);
+			}
+		}
+	}
+
+	/**
+	 * A data input stream that wraps a given data input stream and un-registers
+	 * from a given connection-limiting file system
+	 * (via {@link LimitedConnectionsFileSystem#unregisterInputStream(InStream)}
+	 * upon closing.
+	 */
+	private static final class InStream extends FSDataInputStream implements StreamWithTimeout {
+
+		/** The original data input stream to read from. */
+		private final FSDataInputStream originalStream;
+
+		/** The connection-limiting file system to un-register from. */
+		private final LimitedConnectionsFileSystem fs;
+
+		/** An exception with which the stream has been externally closed. */
+		private volatile StreamTimeoutException timeoutException;
+
+		/** The number of bytes written the last time that the {@link #checkNewBytesAndMark(long)}
+		 * method was called. It is important to initialize this with {@code -1} so that the
+		 * first check (0 bytes) always appears to have made progress. */
+		private volatile long lastCheckBytes = -1;
+
+		/** The timestamp when the last inactivity evaluation was made. */
+		private volatile long lastCheckTimestampNanos;
+
+		/** Flag tracking whether the stream was already closed, for proper inactivity tracking. */
+		private AtomicBoolean closed = new AtomicBoolean();
+
+		InStream(
+				FSDataInputStream originalStream,
+				LimitedConnectionsFileSystem fs) {
+
+			this.originalStream = checkNotNull(originalStream);
+			this.fs = checkNotNull(fs);
+		}
+
+		// --- FSDataOutputStream API implementation
+
+		@Override
+		public int read() throws IOException {
+			try {
+				return originalStream.read();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return 0; // silence the compiler
+			}
+		}
+
+		@Override
+		public int read(byte[] b) throws IOException {
+			try {
+				return originalStream.read(b);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return 0; // silence the compiler
+			}
+		}
+
+		@Override
+		public int read(byte[] b, int off, int len) throws IOException {
+			try {
+				return originalStream.read(b, off, len);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return 0; // silence the compiler
+			}
+		}
+
+		@Override
+		public long skip(long n) throws IOException {
+			try {
+				return originalStream.skip(n);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return 0L; // silence the compiler
+			}
+		}
+
+		@Override
+		public int available() throws IOException {
+			try {
+				return originalStream.available();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return 0; // silence the compiler
+			}
+		}
+
+		@Override
+		public void mark(int readlimit) {
+			originalStream.mark(readlimit);
+		}
+
+		@Override
+		public void reset() throws IOException {
+			originalStream.reset();
```

Comment:
Is there a special reason why this is not using `handleIOException(...)` like other methods?

Reply:
Nope, that is simply an oversight. Will fix that...